### ISSUE TITLE
Driver support for processing notices (working towards tracing and logging).

### ISSUE THREAD

**Issue by hsyed** [jackc/pgx#239]
We are building a framework on top of Postgres which has an incremental event processing component. The business logic for the framework is installed as JS modules. I'd like to add support for logging and perhaps tracing to the framework.

I am thinking `raise NOTICE` is the most appropriate vehicle for this. Listen/Notify is asynchronous and thus not suitable for the simple "net/tracing" / context style of logging / tracing. It is suitable for collecting logs without context though, and perhaps if we used opentracing for collecting spans it would work as they can be put into context  after the fact. 

My assumption is that when I `raise NOTICE` on a backend they should be passed through to the frontend in a `NoticeResponse`. I looked at the driver code and the these are currently ignored. I hacked in some code to print out the responses and I don't think I was receiving the notices I was raising. I expect the configuration for `client_min_messages` is at the wrong level for backends started by pgx. I must be doing something wrong !

So assuming my understanding of how `raise NOTICE` works is correct then step one would be to add support to the protocol processing in pgx. 

Next is consuming the notices, there are just so many possibilities, we could associate the notices with traces in `net/context` (#https://github.com/jackc/pgx/issues/218). opentracing is another avenue. 

Perhaps the client should be able to supply some sort of notice processor to the Conn / Connpool.  That minimises  api changes but we would also need a handle to some per transaction / per connection client state.So perhaps context support (#https://github.com/jackc/pgx/issues/218) is the best vehicle for this.

**Comment by jackc** [comment]
I think support for handling notice in pgx is a good idea. Figuring out the right interface might be tricky though. The `context` lib integration is a possible option. I've got `context` support on the `v3-experimental` branch. However, there is no guarantee a notice will only arrive during query processing.

From the docs: `A frontend must be prepared to accept ErrorResponse and NoticeResponse messages whenever it is expecting any other type of message. See also Section 48.2.6 concerning messages that the backend might generate due to outside events.`

So directly tying it with a context might be challenging. The leaves us with a callback function or buffering the messages to be read later.

Buffering the messages for later processing is probably not reasonable The listen/notify code buffers messages, but the client has to subscribe to receive messages. Notices can arrive unbidden.
So the buffer could grow unbounded.

That leaves us with a callback function. A connection-wide callback function could be limited unless it had some way of knowing if/what query was being processed. Possibly a per-query callback function could be passed in as a `context` variable. Or I'm planning on adding `QueryEx`, `ExecEx`, etc. at some point soon. Maybe a strongly typed callback could be passed there. I'm still not clear on what is idiomatic or correct to pass in a context.

As far as `client_min_messages` I didn't need to do anything to get notices.

I hacked in a `fmt.Println` into the `case noticeResponse:` in `processContextFreeMsg` and the following program does produce output for me.

```go
package main

import (
	"log"

	"github.com/jackc/pgx"
)

func main() {
	config, err := pgx.ParseEnvLibpq()
	if err != nil {
		log.Fatalln(err)
	}

	conn, err := pgx.Connect(config)
	if err != nil {
		log.Fatalln(err)
	}
	defer conn.Close()

	query := `DO language plpgsql $$
BEGIN
   RAISE NOTICE 'hello, world!';
END
$$;`

	_, err = conn.Exec(query)
	if err != nil {
		log.Fatal(err)
	}

}
```
